File: smart-chef-ui/src/shared/product.ts
import { ProductCategoryDto } from "./product-category";

export interface ProductDTO {
  id: string;
  createdAt: Date;
  updatedAt: Date;
  gtin: string;
  name: string;
  description?: string;
  amount: number;
  unit: UNITS;
}

export interface ProductWithCategoryDTO extends ProductDTO {
  category: ProductCategoryDto;
}

export interface CreateProductDto {
  gtin: string;
  name: string;
  amount: number;
  unit: string;
  manufacturer: string;
  description?: string;
  category_id: string;
}

export enum UNITS {
  GRAM = "GRAM",
  KILOGRAM = "KILOGRAM",
  LITER = "LITER",
  MILLILITER = "MILLILITER",
  PIECE = "PIECE",
}

File: smart-chef-api/src/smartchef/serializers/product.py
from asyncore import read
from unicodedata import category
from ..models import Product, ProductManufacturer, ProductCategory
from rest_framework import serializers
from .resource import ResourceSerializer
from .productCategory import ProductCategorySerializer


class ProductSerializer(ResourceSerializer):
    # Override the create method to create the manufacturere if it doesn't exist
    def create(self, validated_data):
        manufacturer_name = validated_data.pop('manufacturer')
        manufacturer, _ = ProductManufacturer.objects.get_or_create(
            name=manufacturer_name)
        validated_data['manufacturer'] = manufacturer

        return super().create(validated_data)

    manufacturer = serializers.CharField(required=False, write_only=True)
    category = ProductCategorySerializer(read_only=True)
    category_id = serializers.UUIDField(write_only=True)

    class Meta:
        model = Product
        fields = ('id', 'createdAt', 'updatedAt', 'gtin',
                  'name', 'description', 'amount', 'unit', 'manufacturer', 'category', 'category_id')
